rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1614 @@
+---
```

Comment:
I've noticed that lists of attrsets look kinda bad for example 
https://github.com/NixOS/nixpkgs/blob/75f9d4f7cf99764b304d60fc6c8963ce565ed2f5/doc/default.nix#L10-L29
turns into 
https://github.com/piegamesde/nixpkgs/blob/e345d638168c066fae7e5bcfee414eae919712b1/doc/default.nix#L13-L79
```diff
     libsets = [
-      { name = "asserts"; description = "assertion functions"; }
-      { name = "attrsets"; description = "attribute set functions"; }
-      { name = "strings"; description = "string manipulation functions"; }
-      { name = "versions"; description = "version string functions"; }
-      { name = "trivial"; description = "miscellaneous functions"; }
-      { name = "fixedPoints"; baseName = "fixed-points"; description = "explicit recursion functions"; }
-      { name = "lists"; description = "list manipulation functions"; }
-      { name = "debug"; description = "debugging functions"; }
-      { name = "options"; description = "NixOS / nixpkgs option handling"; }
-      { name = "path"; description = "path functions"; }
-      { name = "filesystem"; description = "filesystem functions"; }
-      { name = "fileset"; description = "file set functions"; }
-      { name = "sources"; description = "source filtering functions"; }
-      { name = "cli"; description = "command-line serialization functions"; }
-      { name = "gvariant"; description = "GVariant formatted string serialization functions"; }
-      { name = "customisation"; description = "Functions to customise (derivation-related) functions, derivatons, or attribute sets"; }
-      { name = "meta"; description = "functions for derivation metadata"; }
-      { name = "derivations"; description = "miscellaneous derivation-specific functions"; }
+      {
+        name = "asserts";
+        description = "assertion functions";
+      }
+      {
+        name = "attrsets";
+        description = "attribute set functions";
+      }
+      {
+        name = "strings";
+        description = "string manipulation functions";
+      }
+      {
+        name = "versions";
+        description = "version string functions";
+      }
+      {
+        name = "trivial";
+        description = "miscellaneous functions";
+      }
+      {
+        name = "fixedPoints";
+        baseName = "fixed-points";
+        description = "explicit recursion functions";
+      }
+      {
+        name = "lists";
+        description = "list manipulation functions";
+      }
+      {
+        name = "debug";
+        description = "debugging functions";
+      }
+      {
+        name = "options";
+        description = "NixOS / nixpkgs option handling";
+      }
+      {
+        name = "path";
+        description = "path functions";
+      }
+      {
+        name = "filesystem";
+        description = "filesystem functions";
+      }
+      {
+        name = "fileset";
+        description = "file set functions";
+      }
+      {
+        name = "sources";
+        description = "source filtering functions";
+      }
+      {
+        name = "cli";
+        description = "command-line serialization functions";
+      }
+      {
+        name = "gvariant";
+        description = "GVariant formatted string serialization functions";
+      }
+      {
+        name = "customisation";
+        description = "Functions to customise (derivation-related) functions, derivatons, or attribute sets";
+      }
+      {
+        name = "meta";
+        description = "functions for derivation metadata";
+      }
+      {
+        name = "derivations";
+        description = "miscellaneous derivation-specific functions";
+      }
     ];
```
which uses a lot more vertical space and also looks kinda bad (to me).

I don't really have a solution for that tho.

maybe:
```nix
[
  { name = "asserts";
    description = "assertion functions"; }
  { name = "attrsets";
    description = "attribute set functions"; }
]
```
but i think that doesn't fit at all with the general style of the rfc.

Any ideas or thoughts on that?

Reply:
While I'd agree that the re-format reduces readability, not sure how to make this a consistent rule.

Reply:
This is a similar to [my comment about lists](https://github.com/NixOS/rfcs/pull/166#discussion_r1480439213).

I'd therefore also generalise this rule to mean that you MAY have any number of elements in a list or attrset as long as it fits the line limit but you SHOULD expand it when the addition of more elements in the future is likely.

For the formatter, it'd have to take previous formatting into account. For transitioning a non-compliant code-base, it'd be good to have a flag for a stricter mode which forces all MAYs to SHOULDs. Humans can then undo that for the specific edge cases where single line formatting is just better.

Reply:
No matter how we ultimately go about it, I'd like for these downsides and trade-off discussion to be added to the RFC text.

Reply:
I don't see any way to make a rule which would improve the formatting of this code example without making other cases worse.

And given that boilerplate is rarely pretty in nature and should generally be avoided, I am okay with expanding the items here.

We already amended the RFC to allow implementations to preserve the input formatting in limited circumstances. I think that the advantages and downsides of our approach have been sufficiently discussed on a general level in the RFC text, and that there is no need to rehash them for every example somebody finds.

Reply:
The current rules would allow persisting the original formatting, since it only says "few enough":

https://github.com/NixOS/rfcs/blob/aaf1613f51abb97be149a5f429634a25277c2327/rfcs/0166-nix-formatting.md?plain=1#L1403

So while the current implementation doesn't do that, we could easily change this in the future.